rfcs/20231108-gpt-quantization/README.md

```diff
@@ -0,0 +1,136 @@
+   - Zero points are per-output and per-input channels with groups along input channels, data type is u8/s8
+
+# 2. Implications to oneDNN
+
+There are 3 new quantization features that can't be supported by the current oneDNN API:
+- scales and zero points data type. Currently oneDNN defines scales as f32 data and zero points as s32 data.
+- scales and zero points groups support. Currently scales and zero points are set using a mask where each
+  bit represents if for a given dimension multiple or a single scale/zero point is used.
+- Weights quantization applied to primitives with non-integer activations. This part is covered by a separate RFC: [[#4]][4]
+
+# 3. Proposal
+
+## Scales and zero points data type support
+
+### Option a (recommended)
+
+Extend scales and zero points setting API with a parameter for data type. This additional information will allow kernels to generate
+the best code for a given data type.
+
+```cpp
+// C API
+status_t dnnl_primitive_attr_set_scales(
+        primitive_attr_t *attr, int arg, ..., dnnl_data_type_t data_type);
+
+// C++ API
+void set_scales_mask(int arg, ..., data_type_t data_type);
+
+// An example
+primitive_attr attr;
+attr.set_scales(DNNL_ARG_DST, ..., dt::f16);
+...
+auto dst_scales_md = memory::desc({1}, dt::f16, tag::x);
+auto dst_scales = memory(dst_scales_md, engine);
+...
+p.execute(s, {{DNNL_ARG_ATTR_SCALES | DNNL_ARG_DST, dst_scales}, ...});
+```
+
+### Option b
+
+Keep data types implicit during scales and zero points setup, but detect the data type from the memory objects passed at execution.
+
+```cpp
+// An example
+primitive_attr attr;
+attr.set_scales(DNNL_ARG_DST, ...);
+...
+auto dst_scales_md = memory::desc({1}, dt::f16, tag::x);
+auto dst_scales = memory(dst_scales_md, engine);
+...
+p.execute(s, {{DNNL_ARG_ATTR_SCALES | DNNL_ARG_DST, dst_scales} ...});
+```
+
+## Scales and zero points groups support
+
+## Option a (recommended)
+
+Extend scales and zero points setting API with a parameter for groups.
+
+```cpp
+// C API
+status_t dnnl_primitive_attr_set_scales(primitive_attr_t *attr, int arg,
+        int mask, int ndims, const dims_t groups, ...);
+
+// C++ API
+void set_scales_mask(int arg, int mask, dims groups, ...);
+```
+
+## Option b
+
+Extend scales and zero points setting API with a parameter for dims.
+This option is implemented in the fork of oneDNN in OpenVino side.
+The downside of this option is that the dimensions of a primitive are now part of attributes.
+As a result the same attributes can't be reused for multiple primitives or pasted as is into
+nested primitives that compute a part of dimensions (for example, brgemm-based primitives). 
+
+```cpp
+// C API
+status_t dnnl_primitive_attr_set_scales(primitive_attr_t *attr, int arg,
+        int mask, int ndims, const dims_t dims, ...);
+// C++ API
+void set_scales_mask(int arg, int mask, dims dims, ...);
+```
+
```

Comment:
I guess the option is incomplete here. We need to define the broadcasting behavior when scales dimension differs from the dimension it maps to.